pkg/operator/ceph/object/cosi/controller.go

```diff
@@ -0,0 +1,287 @@
+	opcontroller "github.com/rook/rook/pkg/operator/ceph/controller"
+	"github.com/rook/rook/pkg/operator/ceph/reporting"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	packageName                = "ceph-cosi"
+	controllerName             = "ceph-cosi-controller"
+	CephCOSIDriverName  = "ceph-cosi-driver"
+	COSISideCarName            = "objectstorage-provisioner-sidecar"
+	CosiControllerDeployment   = "objectstorage-controller"
+	CosiControllerNamespace    = "default"
+	DefaultCephCOSIDriverImage = "quay.io/jthottan/ceph-cosi-driver:v0.1"
+	DefaultImagePullPolicy     = "IfNotPresent"
+	DefaultCOSISideCarImage    = "gcr.io/k8s-staging-sig-storage/objectstorage-sidecar/objectstorage-sidecar:v20221117-v0.1.0-22-g0e67387"
+	cosiSocketMountPath        = "/var/lib/cosi"
+	DefaultServiceAccountName  = "objectstorage-provisioner-sa"
+	cosiSocketVolumeName       = "socket"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", controllerName)
+
+// ReconcileCephCOSIDriver reconciles the Ceph COSI Driver
+type ReconcileCephCOSIDriver struct {
+	client           client.Client
+	context          *clusterd.Context
+	opManagerContext context.Context
+	recorder         record.EventRecorder
+}
+
+// Add creates a new CephCOSIDriver Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, context *clusterd.Context, opManagerContext context.Context, opConfig opcontroller.OperatorConfig) error {
+	return add(mgr, newReconciler(mgr, context, opManagerContext))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, context *clusterd.Context, opManagerContext context.Context) reconcile.Reconciler {
+	return &ReconcileCephCOSIDriver{
+		client:           mgr.GetClient(),
+		context:          context,
+		opManagerContext: opManagerContext,
+		recorder:         mgr.GetEventRecorderFor(controllerName),
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	controller, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	logger.Info("successfully started")
+	// Watch for changes to CephCOSIDriver
+	err = controller.Watch(&source.Kind{Type: &cephv1.CephCOSIDriver{}}, &handler.EnqueueRequestForObject{})
```

Comment:
also, rebase your branch since we have moved to controller runtime 0.15.0 it brings some changes on watching resources, example here https://github.com/rook/rook/blob/1d3f99bc1455520cb1300600abb8b789f4d71a64/pkg/operator/ceph/cluster/controller.go#L133-L143